tree/ntuple/v7/inc/ROOT/RNTupleProcessor.hxx

```diff
@@ -0,0 +1,194 @@
+/// is welcome!
+
+/*************************************************************************
+ * Copyright (C) 1995-2024, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RNTupleProcessor
+#define ROOT7_RNTupleProcessor
+
+#include <ROOT/REntry.hxx>
+#include <ROOT/RError.hxx>
+#include <ROOT/RNTupleDescriptor.hxx>
+#include <ROOT/RNTupleModel.hxx>
+#include <ROOT/RNTupleUtil.hxx>
+#include <ROOT/RPageStorage.hxx>
+#include <ROOT/RSpan.hxx>
+
+#include <memory>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace ROOT {
+namespace Experimental {
+
+struct RNTupleSourceSpec {
+   std::string fName;
+   std::string fStorage;
+
+   RNTupleSourceSpec() = default;
+   RNTupleSourceSpec(std::string_view n, std::string_view s) : fName(n), fStorage(s) {}
+};
+
+namespace Internal {
+class RNTupleProcessorContext {
+private:
+   struct RFieldContext {
+      std::unique_ptr<ROOT::Experimental::RFieldBase> fProtoField;
+      std::unique_ptr<ROOT::Experimental::RFieldBase> fConcreteField;
+      REntry::RFieldToken fToken;
+      std::shared_ptr<void> fValuePtr;
+
+      RFieldContext(std::unique_ptr<ROOT::Experimental::RFieldBase> protoField, REntry::RFieldToken token)
+         : fProtoField(std::move(protoField)), fToken(token)
+      {
+      }
+
+      void ResetConcreteField() { fConcreteField = nullptr; }
+      void SetConcreteField() { fConcreteField = fProtoField->Clone(fProtoField->GetFieldName()); }
+   };
+
+   std::unique_ptr<RNTupleModel> fModel;
+   std::unique_ptr<REntry> fEntry; ///< The entry is based on the first page source
+   std::unique_ptr<Internal::RPageSource> fPageSource;
+   std::vector<RFieldContext> fFieldContexts;
+
+public:
+   RNTupleProcessorContext(const RNTupleSourceSpec &ntuple);
+
+   /////////////////////////////////////////////////////////////////////////////
+   /// \brief Initializes the processor context
+   ///
+   /// Recursively creates so-called 'protofields' based on the model of the first page source, which will serve as
+   /// a basis for all subsequent RNTuples to be processed.
+   void Initialize();
+
+   /////////////////////////////////////////////////////////////////////////////
+   /// \brief Updates the processor context
+   ///
+   /// \param[in] ntuple The RNTuple to connect for processing
+   ///
+   /// \return The number of entries in the newly-connected RNTuple
+   ///
+   /// Creates and attaches new page source for the specified RNTuple, and connects the fields that are known by
+   /// the processor context to it.
+   NTupleSize_t Update(const RNTupleSourceSpec &ntuple);
```

Comment:
Would `Add` or `Append` be better method names?

Reply:
I don't think those names would be completely accurate, because it replaces the previous page source instead of just adding one. I've renamed it now to `ConnectNTuple` as it also connects the fields to the new ntuple besides just attaching the page source. Of course other suggestions are also welcome!